File: code/class-03-encapsulation/bank_account.py
"""
bank_account.py

Class 03 – Encapsulation

Demonstrates:
  - Private attributes (name-mangled with __)
  - Public getters (properties) and controlled mutators
  - Validation logic inside methods
  - Read-only attribute via property without setter
"""


class BankAccount:
    """A bank account with controlled access to balance."""

    def __init__(self, owner: str, initial_balance: float = 0.0) -> None:
        if initial_balance < 0:
            raise ValueError("Initial balance cannot be negative.")
        self.__owner             = owner          # private
        self.__balance           = initial_balance  # private
        self.__transaction_count = 0               # private

    # ── Read-only properties ─────────────────────────────────────────────────

    @property
    def owner(self) -> str:
        return self.__owner

    @property
    def balance(self) -> float:
        return self.__balance

    @property
    def transaction_count(self) -> int:
        return self.__transaction_count

    # ── Controlled mutators ───────────────────────────────────────────────────

    def deposit(self, amount: float) -> None:
        if amount <= 0:
            print("Deposit amount must be positive.")
            return
        self.__balance += amount
        self.__transaction_count += 1
        print(f"Deposited ${amount:,.2f}. New balance: ${self.__balance:,.2f}")

    def withdraw(self, amount: float) -> bool:
        if amount <= 0:
            print("Withdrawal amount must be positive.")
            return False
        if amount > self.__balance:
            print("Insufficient funds.")
            return False
        self.__balance -= amount
        self.__transaction_count += 1
        print(f"Withdrew ${amount:,.2f}. New balance: ${self.__balance:,.2f}")
        return True

    def __str__(self) -> str:
        return (
            f"Account[owner={self.__owner}, "
            f"balance=${self.__balance:,.2f}, "
            f"transactions={self.__transaction_count}]"
        )


# ── Demo ─────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    account = BankAccount("Alice", 1_000.00)
    print(account)

    account.deposit(500.00)
    account.withdraw(200.00)
    account.withdraw(2_000.00)   # should fail

    print(account)

    # Direct access to __balance is blocked by name mangling
    # account.__balance = 999  # This would create a NEW attribute, NOT modify the private one

File: code/class-03-encapsulation/BankAccount.java
/**
 * BankAccount.java
 *
 * Class 03 – Encapsulation
 *
 * Demonstrates:
 *   - Private fields protected from direct access
 *   - Public getters and controlled mutators
 *   - Validation logic inside methods
 *   - Immutable helper value object (Money)
 */
public class BankAccount {

    private final String owner;
    private double balance;
    private int    transactionCount;

    public BankAccount(String owner, double initialBalance) {
        if (initialBalance < 0) {
            throw new IllegalArgumentException("Initial balance cannot be negative.");
        }
        this.owner            = owner;
        this.balance          = initialBalance;
        this.transactionCount = 0;
    }

    // --- Getters (read-only access) ---
    public String getOwner()            { return owner; }
    public double getBalance()          { return balance; }
    public int    getTransactionCount() { return transactionCount; }

    // --- Controlled mutators ---
    public void deposit(double amount) {
        if (amount <= 0) {
            System.out.println("Deposit amount must be positive.");
            return;
        }
        balance += amount;
        transactionCount++;
        System.out.printf("Deposited $%.2f. New balance: $%.2f%n", amount, balance);
    }

    public boolean withdraw(double amount) {
        if (amount <= 0) {
            System.out.println("Withdrawal amount must be positive.");
            return false;
        }
        if (amount > balance) {
            System.out.println("Insufficient funds.");
            return false;
        }
        balance -= amount;
        transactionCount++;
        System.out.printf("Withdrew $%.2f. New balance: $%.2f%n", amount, balance);
        return true;
    }

    @Override
    public String toString() {
        return String.format("Account[owner=%s, balance=%.2f, tx=%d]",
                owner, balance, transactionCount);
    }

    // ----- Demo -----
    public static void main(String[] args) {
        BankAccount account = new BankAccount("Alice", 1_000.00);
        System.out.println(account);

        account.deposit(500.00);
        account.withdraw(200.00);
        account.withdraw(2_000.00); // should fail

        System.out.println(account);
    }
}
